### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by sindresorhus** [SamVerschueren/gh-lint-brainstorm#3]
Just for some general discussion.

**Comment by sindresorhus** [comment]
I actually created something similar to this 2 years ago, but never had the time to finish it... It's been lying dormant in `~/dev` since then. I've just uploaded it to GitHub in case you find some inspiration in it or not: https://github.com/sindresorhus/npmlint

**Comment by SamVerschueren** [comment]
@sindresorhus Thanks! Mine is just a quick-and-dirty implementation in one file just to see what I can do.

**Comment by jvandemo** [comment]
How about:
- `npm-lint`: library repo name
- `npm-lint-cli`: CLI repo name
- `npm-lint`: command (installed by CLI)

Still looks to be available:

```
'npm-lint' is not in the npm registry.
```

Thanks for making this, looks awesome!

**Comment by SamVerschueren** [comment]
@jvandemo This is not identical to `npm-lint` as `npm-lint` checks the local package (if I'm not mistaken). My tool is meant to iterate over all your github repositories so it's easy to lint packages created a long time ago.

**Comment by SamVerschueren** [comment]
But I agree it would be nice to run it locally as well :).

**Comment by jvandemo** [comment]
@SamVerschueren Oh, sorry, totally misunderstood, that's even more awesome!

Then how about:
- `gh-lint`: library repo name
- `gh-lint-cli`: CLI repo name
- `gh-lint`: command (installed by CLI)

```
'gh-lint' is not in the npm registry.
```

Would you be able to lint other people's repositories as well?

This could be useful if I wanted to lint a repository in our organization that is created by someone else, if that makes sense.

Keep up the good work! :+1:

**Comment by SamVerschueren** [comment]
> Would you be able to lint other people's repositories as well?

Absolutely! I started this project because I contribute to @sindresorhus his repositories (add XO, ES2015ify tests, promisify libs). He has a lot of repositories, don't want to check them all by hand :D. Would be usefully for beginner OSS contributors as well to see what repositories can be improved (quite easily).

**Comment by sindresorhus** [comment]
I like the name `gh-lint`.

**Comment by jvandemo** [comment]
Would be really cool if it supported both local and remote. Then maybe `npm-lint` would make more sense:

``` bash
// lint current directory
$ npm-lint

// lint project directory
$ npm-lint project

// lint github repo
$ npm-lint https://github.com/sindresorhus/pageres

// lint branch of github repo
$ npm-lint https://github.com/sindresorhus/pageres/tree/coverage
```

However, maybe that would go against keeping modules as small as possible.

Maybe both can be implemented where `gh-lint` would use `npm-lint` behind the scenes.

Can't wait to start using it already! ;-)

**Comment by sindresorhus** [comment]
I don't see why `gh-lint` couldn't work both locally and remotely.

**Comment by SamVerschueren** [comment]
The basic idea was to lint all the repositories (if no argument) is provided. Something like this:

```
// Lint ALL repositories
$ gh-lint -u sindresorhus

// Only lint pageres repository
$ gh-lint pageres -u sindresorhus

// Lint local pageres
$ gh-lint ./pageres
```

_(Just a simple example, everything is pretty raw at the moment)_

> Can't wait to start using it already! ;-)

Good to see people are enthusiastic about the project :)

**Comment by jvandemo** [comment]
> I don't see why gh-lint couldn't work both locally and remotely.

True that!

Maybe one counterargument would be that `gh-lint` would not be a name that people look for in the npm registry when they have a project that is not stored on GitHub (or even not a git repo at all)?

Other than that, I love the name too! :+1:

**Comment by sindresorhus** [comment]
Something not on GitHub is an edge-case today.

**Comment by parro-it** [comment]
Today I quickly lloked at [sindresorhus/npmlint](https://github.com/sindresorhus/npmlint). 
I really like the possibility of having configurable rules, maybe using presets like Babel 6 is doing...

Or even better, I would want to publish a module named `parro-it/gh-lint-preset` and that will become the default preset to use for all my repos.

**Comment by SamVerschueren** [comment]
Hi guys, been a while. Want to give you a quick update on how the package is going.

I think I figured out the core of the package. But feedback is much appreciated, as usual! I have some rules defined as you can see [here](https://github.com/SamVerschueren/gh-lint/tree/master/rules). Feel free to give suggestions on names/implementations/error messages of rules.

I have no cli at the moment, but linting can be done like this

``` js
const linter = require('./index.js');

linter('SamVerschueren/gh-lint', {
    rules: {
        'editorconfig': 2,
        'use-promise': 2,
        'use-travis': 1,
        'travis-build-status': 1,
        'valid-properties': [2, 'version', 'license'],
        'valid-version': 2,
        'license-mit': 2
    }
}).then(errors => {
    console.log(errors);
})
```

where `1` is a `warning` and `2` is an `error`. 

Local modules can be linted as well. And that's why I wanted to reconsider the name of the linter. It started out with the idea of linting GitHub repositories. But I feel that the module is more a project structure linter as it looks at the existance of files, configuration, etc. Plugins are not implemented yet, but when they do, it would be possible for instance to create a `my-company-react` plugin that validates the project structure: "Are the views placed in the correct location?" "Do we have a file x, y and z and the correct place". Like to hear thoughts of others on this. And off course, suggestions on the current and maybe new name?

**Comment by sindresorhus** [comment]
@SamVerschueren Looks good!

> where 1 is a warning and 2 is an error.

ESLint recently changed to `'off'` | `'warn'` | `'error'`. I think that's a lot more user-friendly.

## 

I don't think rules should throw/reject a normal Error. That makes it impossible to differentiate rule violation from actual errors. Instead either use a custom error type or even better, just return a result object.

## 

I would like to also see a property to enforce order of properties in package.json.

A `--fix` option for that would be amazing.

## 

Use arrow functions. You're already targeting Node.js 4.

## 

I agree with the need of a name change and your proposed project direction. And to be honest, I would only use it locally. Find a name that is short.

**Comment by SamVerschueren** [comment]
> I don't think rules should throw/reject a normal Error. That makes it impossible to differentiate rule violation from actual errors. Instead either use a custom error type or even better, just return a result object.

Fixed

> Use arrow functions. You're already targeting Node.js 4.

Fixed

> I would like to also see a property to enforce order of properties in package.json.

Just added a [rule](https://github.com/SamVerschueren/gh-lint/blob/master/rules/pkg-schema.js) that will validate the `package.json` based upon a schema, so that's a start. Any suggestions on the order of the properties? Should these be hardcoded or should it be an arguments list?

---

Yes about the name, I was thinking about
- `modulint`
- `modlint`
- `pkglint`
- `prolint`
- ...

It's hard to choose because once released, it's almost impossible to change it to something else.

**Comment by sindresorhus** [comment]
> Just added a rule that will validate the package.json based upon a schema, so that's a start. Any suggestions on the order of the properties? Should these be hardcoded or should it be an arguments list?

Yes. If you build an array of all the possible properties, I'll do a PR suggestion the default order. Should also be configurable. Non-official properties, like `xo` and `ava`), should be put at the bottom.

> It's hard to choose because once released, it's almost impossible to change it to something else.

Then do something generic and fun. Like I did with `meow`.

**Comment by jvandemo** [comment]
> Then do something generic and fun. Like I did with meow.

How about `clinton`?

It has the words `cli` and `lint` in it and it's easy to remember.

Oh, and it's (surprisingly) still available on npm.

**Comment by SamVerschueren** [comment]
Oh, I like `clinton`! Thanks for the suggestion @jvandemo :).

**Comment by jvandemo** [comment]
Welcome @SamVerschueren, thank **you** for making this real!

**Comment by SamVerschueren** [comment]
Released [clinton@v0.1.0](https://www.npmjs.com/package/clinton). Feel free to test it out, post suggestions for improvements or new rules. Anything really! Learned from the past that you can only get the most out of a module/library if you just use it and look for bad parts (and good parts) and improve. I guess that's also how `AVA` was started :).

**Comment by jvandemo** [comment]
@SamVerschueren &mdash; Cool, just installed it and works nicely! 👍

**Comment by sindresorhus** [comment]
Didn't work for me:

```
~/dev/ava master
❯ clinton --version
0.1.0

~/dev/ava master
❯ clinton          
Error: No rules found
    at /usr/local/lib/node_modules/clinton/index.js:42:11
```

You're probably using `process.cwd()` instead of `__dirname` somewhere.

**Comment by SamVerschueren** [comment]
@sindresorhus I didn't define a default set of rules (yet). Does ESLint expose a default set of rules? It seems very opinionated. That's why you created `XO`, right? 

I just released `0.2.0` which offers the `AVA` and `XO` rules + some extras. Try adding the following config to `package.json`

``` json
"clinton": {
  "rules": {
    "xo": ["error", "*"],
    "license": ["error", "ISC"]
  }
}
```

Run `clinton` again.

It will fail because the license is not `ISC`. Change it to `MIT` and it will succeed.

Might have to work on my library error messages though ;).

**Comment by sindresorhus** [comment]
Oh, I misunderstood the intention of this project. I thought you were going for something opinionated like XO, not ESLint.

**Comment by SamVerschueren** [comment]
I can still go that direction. I didn't do it yet because I wanted to focus on the core first. Actually it would be pretty easy to let people overwrite the default set of rules with a config as shown above. This way the defaults are still opinionated, but people can create their own config if they want. It's a win-win :).

**Comment by SamVerschueren** [comment]
@sindresorhus Released `clinton@0.3.0` with a default set of [rules](https://github.com/SamVerschueren/clinton/blob/master/config.js) :).

**Comment by sindresorhus** [comment]
Cool. Will run it on some projects.

**Comment by sindresorhus** [comment]
> error    Found property 'scripts' but expected 'engines'. (pkg-property-order)

(from running it on https://github.com/sindresorhus/load-grunt-tasks)

The error message could be more helpful. See how it's handled by the [`order` rule](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md) in `eslint-plugin-import`.

**Comment by SamVerschueren** [comment]
Good suggestion!

**Comment by sindresorhus** [comment]
@SamVerschueren Also show file and line:column whenever possible.

I got `error    Use promises instead of callbacks (no-callback)` and have no idea where it's violating.

**Comment by SamVerschueren** [comment]
@sindresorhus what package?

**Comment by sindresorhus** [comment]
@SamVerschueren Cmd+F the gist I sent you on Hangouts.
